From 94d0052d2eacb23c98f05baf016170e9c92f5f61 Mon Sep 17 00:00:00 2001
From: Rameshkumar Sundaram <quic_ramess@quicinc.com>
Date: Thu, 20 Jun 2024 17:13:54 +0530
Subject: [PATCH 3/4] hostapd: Add CLI support for ML link addition

Add new CLI command mld_add_link to add new link
Sample command:
hostapd_cli -i wlan0 mld_add_link bss_config=phy0:6g.conf

Signed-off-by: Rameshkumar Sundaram <quic_ramess@quicinc.com>
Signed-off-by: Sowmiya Sree Elavalagan <quic_ssreeela@quicinc.com>
---
 hostapd/ctrl_iface.c  |  5 +++++
 hostapd/hostapd_cli.c | 11 +++++++++++
 2 files changed, 16 insertions(+)

diff --git a/hostapd/ctrl_iface.c b/hostapd/ctrl_iface.c
index 33c18ba..dc86c73 100644
--- a/hostapd/ctrl_iface.c
+++ b/hostapd/ctrl_iface.c
@@ -94,6 +94,8 @@
 static void hostapd_ctrl_iface_send(struct hostapd_data *hapd, int level,
 				    enum wpa_msg_type type,
 				    const char *buf, size_t len);
+static int hostapd_ctrl_iface_add(struct hapd_interfaces *interfaces,
+				  char *buf);
 
 static char *reload_opts = NULL;
 
@@ -5000,6 +5002,9 @@ static int hostapd_ctrl_iface_receive_process(struct hostapd_data *hapd,
 							  reply_size);
 #endif /* ANDROID */
 #ifdef CONFIG_IEEE80211BE
+	} else if (os_strncmp(buf, "MLD_ADD_LINK ", 13) == 0) {
+		if (hostapd_ctrl_iface_add(hapd->iface->interfaces, buf + 13))
+			reply_len = -1;
 #ifdef CONFIG_TESTING_OPTIONS
 	} else if (os_strncmp(buf, "LINK_REMOVE ", 12) == 0) {
 		if (hostapd_ctrl_iface_link_remove(hapd, buf + 12,
diff --git a/hostapd/hostapd_cli.c b/hostapd/hostapd_cli.c
index 7632ea8..8b589bc 100644
--- a/hostapd/hostapd_cli.c
+++ b/hostapd/hostapd_cli.c
@@ -1470,6 +1470,13 @@ static int hostapd_cli_cmd_req_range(struct wpa_ctrl *ctrl, int argc,
 	return hostapd_cli_cmd(ctrl, "REQ_RANGE", 4, argc, argv);
 }
 
+#ifdef CONFIG_IEEE80211BE
+static int hostapd_cli_cmd_mld_add_link(struct wpa_ctrl *ctrl, int argc,
+					char *argv[])
+{
+	return hostapd_cli_cmd(ctrl, "MLD_ADD_LINK", 1, argc, argv);
+}
+#endif
 
 static int hostapd_cli_cmd_driver_flags(struct wpa_ctrl *ctrl, int argc,
 					char *argv[])
@@ -1896,6 +1903,10 @@ static const struct hostapd_cli_cmd hostapd_cli_commands[] = {
 	{ "driver", hostapd_cli_cmd_driver, NULL,
 	  "<driver sub command> [<hex formatted data>] = send driver command data" },
 #endif /* ANDROID */
+#ifdef CONFIG_IEEE80211BE
+	{ "mld_add_link", hostapd_cli_cmd_mld_add_link, NULL,
+	"<config_file_location>" },
+#endif
 	{ NULL, NULL, NULL, NULL }
 };
 
-- 
2.34.1

